Machine learning segmentation

leonardosepulveda opened the pull request:
I changed the name of nuclei_segmentation_cv2 to "machine_learning_segmentation". I will keep the cv2 segmentation capability and add a new class to allow the implementation of machine learning approaches. I will implement ilastik, U-nets and cellpose to begin with. The user will need to provide a trained model that can be used to segment either nuclei or cells. I will add a test function to compare the results of all available methods in a few FOVs.

leonardosepulveda committed:
Adding _get_membrane_mask and _get_nuclei_mask functions to WatershedSegmentNucleiCV2 class

leonardosepulveda committed:
changed formatting to comply to pep8

leonardosepulveda committed:
add borderMask to _get_nuclei_mask function

leonardosepulveda committed:
added method _generate_markers to WatershedSegmentNucleiCV2

leonardosepulveda committed:
added method _generate_markers to WatershedSegmentNucleiCV2

leonardosepulveda committed:
added method _apply_watershed

leonardosepulveda committed:
added method _convert_grayscale_to_rgb

leonardosepulveda committed:
modify _apply_watershed method

leonardosepulveda committed:
adding _combine_watershed_z_positions method, starting by pseudocode

leonardosepulveda committed:
expanding _combine_watershed_z_positions and adding _get_overlapping_nuclei method

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
added _read_and_filter_image_stack to WatershedSegmentNucleiCV2

leonardosepulveda committed:
reorganized Watershed...CV2 to remove repeated method calls

leonardosepulveda committed:
cleaned watershed

leonardosepulveda committed:
cleaned watershed

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
fixing invalid syntax

leonardosepulveda committed:
correct skimage function names

leonardosepulveda committed:
correct skimage function names

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
correct skimage function names

leonardosepulveda committed:
changed Image dimension order to fit MERlin's

leonardosepulveda committed:
changed Image dimension order to fit MERlin's

leonardosepulveda committed:
change variable name

leonardosepulveda committed:
added missing self. in method call

leonardosepulveda committed:
adding print statements for debugging

leonardosepulveda committed:
adding print statements for debugging

leonardosepulveda committed:
moving utility functions from segment.py to watershed.py

leonardosepulveda committed:
moving utility functions from segment.py to watershed.py

leonardosepulveda committed:
moving utility functions from segment.py to watershed.py

leonardosepulveda committed:
removing utility functions from segment.py

leonardosepulveda committed:
removing utility functions from segment.py

leonardosepulveda committed:
removing utility functions from segment.py

leonardosepulveda committed:
add comments to the header of multiple methods.

leonardosepulveda committed:
add comments to the header of multiple methods.

leonardosepulveda committed:
add comments to the header of multiple methods.

leonardosepulveda committed:
remove commented methods

leonardosepulveda committed:
remove commented methods

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
add george's modifications to dataportal

leonardosepulveda committed:
add george's modifications to dataportal

leonardosepulveda committed:
add george's modifications to dataportal

leonardosepulveda committed:
adding additional modifications to aws

leonardosepulveda committed:
adding additional modifications to aws

leonardosepulveda committed:
fixing indentation

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
added printing for debuging

leonardosepulveda committed:
removing self calls

leonardosepulveda committed:
removing self calls

leonardosepulveda committed:
removing self calls

leonardosepulveda committed:
remove self calls

leonardosepulveda committed:
remove self calls

leonardosepulveda committed:
remove self calls

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
change WatershedSegmentNucleiCV2 to WatershedSegmentCV2, modify accordingly to allow for cytoplasmic or nuclei segmentation

leonardosepulveda committed:
change WatershedSegmentNucleiCV2 to WatershedSegmentCV2, modify accordingly to allow for cytoplasmic or nuclei segmentation

leonardosepulveda committed:
change WatershedSegmentNucleiCV2 to WatershedSegmentCV2, modify accordingly to allow for cytoplasmic or nuclei segmentation

pep8speaks commented:
Hello @leonardosepulveda! Thanks for updating this PR.

* In the file [`merlin/analysis/segment.py`](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/analysis/segment.py):

> [Line 293:72](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/analysis/segment.py#L293): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace







* In the file [`merlin/util/segmentation.py`](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py):

> [Line 139:9](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L139): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 184:22](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L184): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 184:22](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L184): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 185:24](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L185): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 185:24](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L185): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 186:23](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L186): [E261](https://duckduckgo.com/?q=pep8%20E261) at least two spaces before inline comment
> [Line 186:23](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L186): [E262](https://duckduckgo.com/?q=pep8%20E262) inline comment should start with '# '
> [Line 285:59](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L285): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 295:77](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L295): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 297:71](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L297): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 363:49](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L363): [E221](https://duckduckgo.com/?q=pep8%20E221) multiple spaces before operator
> [Line 363:52](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L363): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 363:60](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L363): [E222](https://duckduckgo.com/?q=pep8%20E222) multiple spaces after operator
> [Line 363:81](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L363): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (93 > 80 characters)
> [Line 402:58](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L402): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 513:67](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L513): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 560:61](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L560): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 561:49](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L561): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 562:65](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L562): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 563:49](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/segmentation.py#L563): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

* In the file [`merlin/util/watershed.py`](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/watershed.py):

> [Line 143:46](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/merlin/util/watershed.py#L143): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

* In the file [`test/conftest.py`](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/test/conftest.py):

> [Line 62:81](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/test/conftest.py#L62): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)
> [Line 64:81](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/test/conftest.py#L64): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 80 characters)
> [Line 69:81](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/test/conftest.py#L69): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 80 characters)

* In the file [`test/test_merfish_segmentation_cellpose.py`](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/test/test_merfish_segmentation_cellpose.py):

> [Line 12:81](https://github.com/emanuega/MERlin/blob/8d112f6b82ca027d495174eed2d799934f72e9c4/test/test_merfish_segmentation_cellpose.py#L12): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 80 characters)





##### Comment last updated at 2021-01-08 21:09:37 UTC

merlin-pep8speaks commented:
Hello @leonardosepulveda! Thanks for updating this PR.

* In the file [`merlin/analysis/segment.py`](https://github.com/emanuega/MERlin/blob/91f6ed4e1a15653c61472925f240d988d79f9be1/merlin/analysis/segment.py):

> [Line 141:67](https://github.com/emanuega/MERlin/blob/91f6ed4e1a15653c61472925f240d988d79f9be1/merlin/analysis/segment.py#L141): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 142:72](https://github.com/emanuega/MERlin/blob/91f6ed4e1a15653c61472925f240d988d79f9be1/merlin/analysis/segment.py#L142): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 143:72](https://github.com/emanuega/MERlin/blob/91f6ed4e1a15653c61472925f240d988d79f9be1/merlin/analysis/segment.py#L143): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 153:1](https://github.com/emanuega/MERlin/blob/91f6ed4e1a15653c61472925f240d988d79f9be1/merlin/analysis/segment.py#L153): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

leonardosepulveda committed:
change nuclei to compartment

leonardosepulveda committed:
change nuclei to compartment

leonardosepulveda committed:
change nuclei to compartment

leonardosepulveda committed:
change nuclei to compartment

leonardosepulveda committed:
change nuclei to compartment

leonardosepulveda committed:
change nuclei to compartment

leonardosepulveda committed:
added framework for MachineLearningSegment class

leonardosepulveda committed:
added framework for MachineLearningSegment class

leonardosepulveda committed:
added framework for MachineLearningSegment class

leonardosepulveda committed:
adding mls utils function

leonardosepulveda committed:
adding mls utils function

leonardosepulveda committed:
adding mls utils function

leonardosepulveda committed:
add function definitions to mls.py

leonardosepulveda committed:
add function definitions to mls.py

leonardosepulveda committed:
add function definitions to mls.py

leonardosepulveda committed:
add body of apply_machine_learning_segmentation function

leonardosepulveda committed:
add body of apply_machine_learning_segmentation function

leonardosepulveda committed:
add body of apply_machine_learning_segmentation function

leonardosepulveda committed:
change nuclei to compartment in cv2 segmentation functions

leonardosepulveda committed:
change nuclei to compartment in cv2 segmentation functions

leonardosepulveda committed:
change nuclei to compartment in cv2 segmentation functions

leonardosepulveda committed:
adding edge memebrane calculation to CV2 segmenetation

leonardosepulveda committed:
adding edge memebrane calculation to CV2 segmenetation

leonardosepulveda committed:
adding edge memebrane calculation to CV2 segmenetation

leonardosepulveda committed:
modify get_membrane_mask function

leonardosepulveda committed:
modify get_membrane_mask function

leonardosepulveda committed:
change watershed.py to segmentation.py, put contents of mls.py into segmentation.py

leonardosepulveda committed:
change watershed.py to segmentation.py, put contents of mls.py into segmentation.py

leonardosepulveda committed:
change watershed.py to segmentation.py, put contents of mls.py into segmentation.py

leonardosepulveda committed:
consolidate functions into segment_using_cellpose

leonardosepulveda committed:
consolidate functions into segment_using_cellpose

leonardosepulveda committed:
consolidate functions into segment_using_cellpose

leonardosepulveda committed:
add diameter param to segment_using_cellpose

leonardosepulveda committed:
add diameter param to segment_using_cellpose

leonardosepulveda committed:
put segmentation params into dict

leonardosepulveda committed:
put segmentation params into dict

leonardosepulveda committed:
put segmentation params into dict

leonardosepulveda committed:
add parameters as dict

leonardosepulveda committed:
add parameters as dict

leonardosepulveda committed:
add parameters as dict

leonardosepulveda committed:
change indentation

leonardosepulveda committed:
change indentation

leonardosepulveda committed:
change indentation

leonardosepulveda committed:
return watershed class name to previous state

leonardosepulveda committed:
return watershed class name to previous state

leonardosepulveda committed:
return watershed class name to previous state

leonardosepulveda committed:
add 2D>3D segmentation mask

leonardosepulveda committed:
add 2D>3D segmentation mask

leonardosepulveda committed:
add 2D>3D segmentation mask

leonardosepulveda committed:
changed segmentation output variable name

leonardosepulveda committed:
changed segmentation output variable name

leonardosepulveda committed:
changed segmentation output variable name

leonardosepulveda committed:
changed value of the background mask from 100 to 0

leonardosepulveda committed:
changed value of the background mask from 100 to 0

leonardosepulveda committed:
changed value of the background mask from 100 to 0

leonardosepulveda committed:
mofified botched indentation in combine 2d>3d

leonardosepulveda committed:
modifying tuple output in 2d>3d

leonardosepulveda committed:
modifying tuple output in 2d>3d

leonardosepulveda committed:
modifying tuple output in 2d>3d

leonardosepulveda committed:
modifying tuple output in 2d>3d

leonardosepulveda committed:
made tuple explicit

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
pep8 compliance

leonardosepulveda committed:
added Machine Learning Segment and CV2 to tests

leonardosepulveda committed:
added Machine Learning Segment and CV2 to tests

leonardosepulveda committed:
added Machine Learning Segment and CV2 to tests

leonardosepulveda commented:
I would like to merge this branch to master. I did tests and the new machine learning segmentation using cellpose is working, as well as the watershed CV2 implementation.However, there are some incompatibilities with merlin/utils/dataportal.py. How should I solve that issue?

leonardosepulveda committed:
remove print statements for timing

leonardosepulveda committed:
remove print statements for timing

leonardosepulveda committed:
remove print statements for timing

emanuega commented:
What is the incompatible you're finding with dataportal.py? I prefer for all the raw data access to go through the dataportal  so that  it is easy to switch between AWS, GC, or local storage.

leonardosepulveda committed:
copied dataportal.py from master

leonardosepulveda committed:
copied dataportal.py from master

leonardosepulveda committed:
removed explicit tuple output from function

leonardosepulveda committed:
added explicit tuple output

leonardosepulveda committed:
added explicit tuple output

leonardosepulveda committed:
added explicit tuple output

leonardosepulveda committed:
add cellpose to requirements

leonardosepulveda committed:
add cellpose to requirements

leonardosepulveda committed:
add cellpose to requirements

leonardosepulveda committed:
updated docutils and pillow requirements

leonardosepulveda committed:
updated docutils and pillow requirements

leonardosepulveda committed:
updated docutils and pillow requirements

leonardosepulveda committed:
changed marker bg in ge_cv2_markers

leonardosepulveda committed:
removed docutils requirement to avoid conflict with v017

leonardosepulveda committed:
removed docutils requirement to avoid conflict with v017

leonardosepulveda commented:
I am trying to solve the issues that appear in workflow and ci. So far I understand the output from both so I can go back into the code and make modifications. However, now the build stops in the testing part, but it is unclear to me where the problem is:
```
#!/bin/bash -eo pipefail 
source ~/miniconda/bin/activate root 
conda activate base 
source activate merlin_env 
cd ~/MERlin 
mkdir ~/test-reports 
pytest --cov --cov-report=xml 
 
============================= test session starts ==============================
 
platform linux -- Python 3.6.10, pytest-5.4.3, py-1.9.0, pluggy-0.13.1
rootdir: /home/circleci/MERlin
 plugins: cov-2.10.0
 collecting ... collecting 71 items                                                            collecting 96 items                                                            collected 96 items                                                             
 
test/test_barcode_database.py ..........                                 [ 10%]
 test/test_binary_utils.py ..                                             [ 12%]
test/test_codebook.py ...........                                        [ 23%]
test/test_core.py .....................                                  [ 45%]
test/test_dataorganization.py ..........                                 [ 56%]
 test/test_dataportal.py ............                                     [ 68%]
test/test_dataset.py ...                                                 [ 71%]
 test/test_image_reader.py .                                              [ 72%]
  
Too long with no output (exceeded 10m0s): context deadline exceeded 
```

Any insights on how to debug? other place where I should look?

leonardosepulveda committed:
added input variables to get_membrane_mask

leonardosepulveda committed:
added input variables to get_membrane_mask

leonardosepulveda committed:
added input variables to get_membrane_mask

leonardosepulveda committed:
added additional variable

leonardosepulveda committed:
added additional variable

emanuega commented:
It looks like you are a few commits behind the master and are incorporating some of the changes independently. Can you rebase this branch onto the head of the master branch so we can get all the commits incorporated? (see https://medium.com/@gabriellamedas/git-rebase-and-git-rebase-onto-a6a3f83f9cce). Then we can see if this problem persists.

leonardosepulveda committed:
updated variable name

leonardosepulveda commented:
Thanks for the tip on rebase. It seems that the things started to work well, but the problem comes again with the changes I added to dataportal.py to make it look like the updated branch. See below what I get:
```
[lsepulvedaduran@boslogin03 MERlin]$ git rebase master machine_learning_segmentation
First, rewinding head to replay your work on top of it...
Applying: Adding _get_membrane_mask and _get_nuclei_mask functions to WatershedSegmentNucleiCV2 class
...
Applying: add george's modifications to dataportal
Using index info to reconstruct a base tree...
M       requirements.txt
M       test/test_dataportal.py
Falling back to patching base and 3-way merge...
Auto-merging requirements.txt
Applying: adding additional modifications to aws
Using index info to reconstruct a base tree...
M       merlin/util/dataportal.py
Falling back to patching base and 3-way merge...
Auto-merging merlin/util/dataportal.py
CONFLICT (content): Merge conflict in merlin/util/dataportal.py
Failed to merge in the changes.
Patch failed at 0066 adding additional modifications to aws
The copy of the patch that failed is found in:
   /n/home06/lsepulvedaduran/Software/MERlin/.git/rebase-apply/patch

When you have resolved this problem, run "git rebase --continue".
If you prefer to skip this patch, run "git rebase --skip" instead.
To check out the original branch and stop rebasing, run "git rebase --abort".

[lsepulvedaduran@boslogin03 MERlin]$
```


I looked at the patch 0066 , this is what says:

```
_[lsepulvedaduran@boslogin03 rebase-apply]$ more 0066
From 60acd7246a1a06562dd50a0147ccfb3d810dd35e Mon Sep 17 00:00:00 2001
From: leonardosepulveda <leonardosepulveda@gmail.com>
Date: Sat, 15 Feb 2020 09:38:38 -0500
Subject: adding additional modifications to aws

---
 merlin/util/dataportal.py | 37 ++++++++++++++++++++++++++++++++++---
 1 file changed, 34 insertions(+), 3 deletions(-)

diff --git a/merlin/util/dataportal.py b/merlin/util/dataportal.py
index e3d0f6e832b69a6491bee876abd741b44844ddb9..9fcd4631bcb603feeeddee777b1f1d9b3f8ffb0e 100644
--- a/merlin/util/dataportal.py
+++ b/merlin/util/dataportal.py
@@ -2,9 +2,11 @@ import os
 import boto3
 import botocore
 from google.cloud import storage
+from google.cloud import exceptions
 from urllib import parse
 from abc import abstractmethod, ABC
 from typing import List
+from time import sleep


 class DataPortal(ABC):
@@ -340,12 +342,41 @@ class GCloudFilePortal(FilePortal):
         return GCloudFilePortal(
             self._exchange_extension(newExtension), self._client)

+    def _error_tolerant_access(self, request, attempts = 5, wait = 60):
+        while attempts > 0:
+            try:
+                file = self._fileHandle.download_as_string()
+                return file
+            except (exceptions.GatewayTimeout, exceptions.ServiceUnavailable):
+                attempts -= 1
+                sleep(60)
+                if attempts == 0:
+                    raise
+
     def read_as_text(self):
-        return self._fileHandle.download_as_string().decode('utf-8')
+         backoffSeries= [1,2,4,8,16,32,64,128,256]
+        for sleepDuration in backoffSeries:
+            try:
+                file = self._fileHandle.download_as_string().decode('utf-8')
+                return file
+            except (exceptions.GatewayTimeout, exceptions.ServiceUnavailable):
+                if sleepDuration == backoffSeries[-1]:
+                    raise
+                else:
+                    sleep(sleepDuration)

     def read_file_bytes(self, startByte, endByte):
-        return self._fileHandle.download_as_string(
-            start=startByte, end=endByte-1)
+        backoffSeries= [1,2,4,8,16,32,64,128,256]
+            for sleepDuration in backoffSeries:
+            try:
+                file = self._fileHandle.download_as_string(start=startByte,
+                                                           end=endByte-1)
+                return file
+            except (exceptions.GatewayTimeout, exceptions.ServiceUnavailable):
+                if sleepDuration == backoffSeries[-1]:
+                    raise
+                else:
+                    sleep(sleepDuration)

     def close(self) -> None:
         pass
--
1.8.3.1_
```

Any idea on how to proceed?

HazenBabcock commented:
If you open `merlin/util/dataportal.py` you should see some notes about merge conflicts, fix these and then you should be able to continue with the rebase.

leonardosepulveda committed:
merge conflict to req..txt, datap..py seg..py

leonardosepulveda commented:
> If you open `merlin/util/dataportal.py` you should see some notes about merge conflicts, fix these and then you should be able to continue with the rebase.

Thanks ... a lot of changes were necessary. For the next time, I will rebase more often while working on a branch.

leonardosepulveda committed:
added pillow and cellpose requirements

leonardosepulveda committed:
added pillow and cellpose requirements

leonardosepulveda commented:
I am still having the same issue as before, the workflow stops in the test part after 10 min in the test_image_reader part. After rebasing, the only thing I did was to add the following two requirements:
```
pillow<=7.0.0
cellpose
```
The pillow requirement comes from the cellpose module. Any idea what can be happening? I do not see any obvious problem with test_image_reader.

emanuega commented:
I think snakemake is hanging. I encountered this in my latest branch and updated the default snakemake parameters. Does https://github.com/emanuega/MERlin/commit/0f3a617bf55c1c108747f4aeb437f14f25d974c6 help?

leonardosepulveda committed:
solve merge conflict

leonardosepulveda committed:
Merge branch 'v0.1.7' into machine_learning_segmentation

leonardosepulveda committed:
remove redundant functions from watershed.py

leonardosepulveda committed:
solve merge conflict

leonardosepulveda committed:
remove redundant tasks

leonardosepulveda committed:
add snakemakeParameters

leonardosepulveda committed:
increase latency_wait

leonardosepulveda commented:
> I think snakemake is hanging. I encountered this in my latest branch and updated the default snakemake parameters. Does [0f3a617](https://github.com/emanuega/MERlin/commit/0f3a617bf55c1c108747f4aeb437f14f25d974c6) help?

I added the modifications to snakemakeParameters in [0f3a617], but the problem persists.

leonardosepulveda committed:
remove MLS and CV2 from merfish test

leonardosepulveda committed:
add test for new segmentation routines

leonardosepulveda commented:
I found the problem. It was in the test_merfish part. I added some tasks for the new segmentation, and the problem seems to come from there. I created independent tests for debugging.

leonardosepulveda commented:
The build run to the end now but the new test files I created cannot be found, this is the complaints

```
FileNotFoundError: [Errno 2] No such file or directory: '/home/circleci/MERlin/test_analysis_parameters/test_analysis_segmentation_cellpose.json'
 
FileNotFoundError: [Errno 2] No such file or directory: '/home/circleci/MERlin/test_analysis_parameters/test_analysis_segmentation_cv2.json'
```

the folder test_analysis_parameters do not exist in my current branch, is there some data reorganization that I am missing?

emanuega commented:
While running the tests it creates this directory and copies the files from the test/auxiliary files directory. You may have to update the base_files function in test/conftest.py to copy your test_analysis_segmentation json configurations into this folder.

leonardosepulveda committed:
add segmentation jsons to conftest

leonardosepulveda committed:
added unique task analysis_name for cellpose and cv2 tests

leonardosepulveda committed:
hard-coded to variable array size in convert_grayscale_to_rgb

leonardosepulveda committed:
added print statement for debogging

leonardosepulveda committed:
change test function names; added fiducial warp and globalign for seg tests

leonardosepulveda committed:
add missing commas to json file

leonardosepulveda committed:
decrease snakemake latency wait

leonardosepulveda commented:
Thanks. Now the tests launch, but they are crashing for a reason I do not understand. I created the new tests by copying and modifying test_merfish.py and test_analysis_parameters.json. The current error is 

```
______________________________ test_cv2_2d_local _______________________________

simple_merfish_data = <merlin.core.dataset.MERFISHDataSet object at 0x7f4636cba978>
     
    @pytest.mark.fullrun
    @pytest.mark.slowtest
    def test_cv2_2d_local(simple_merfish_data):
       	with open(os.sep.join([merlin.ANALYSIS_PARAMETERS_HOME,
                               'test_analysis_segmentation_cv2.json']), 'r') as f:
            snakefilePath = m.generate_analysis_tasks_and_snakefile(
                simple_merfish_data, f)
>           m.run_with_snakemake(simple_merfish_data, snakefilePath, 5)
 
test/test_segmentation_cv2.py:15: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
merlin/merlin.py:194: in run_with_snakemake
    for t in dataSet.get_analysis_tasks()}
 
merlin/merlin.py:194: in <dictcomp>
    for t in dataSet.get_analysis_tasks()}
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _  
 
self = <merlin.core.dataset.MERFISHDataSet object at 0x7f4636cba978>
analysisTaskName = 'RandomNumberParallelAnalysisTask'
    def load_analysis_task(self, analysisTaskName: str) \
            -> analysistask.AnalysisTask:
        loadName = os.sep.join([self.get_task_subdirectory(
            analysisTaskName), 'task.json'])
        print(loadName)
>       with open(loadName, 'r') as inFile:
E       FileNotFoundError: [Errno 2] No such file or directory: '/home/circleci/MERlin/test_analysis/merfish_test/RandomNumberParallelAnalysisTask/tasks/task.json'

merlin/core/dataset.py:620: FileNotFoundError
```

For what I understand that random number task is not called in the text_merfish code, so I do not know why it is complaining. I notice that the test that is run before the segmentation tests is test_plotting, which calls the randomNumber tasks. Any insights?

leonardosepulveda committed:
comment snakemake params

leonardosepulveda committed:
fix comment indentation

leonardosepulveda committed:
change json files for test segmentation

leonardosepulveda committed:
update conftest with new json names

leonardosepulveda committed:
corrected variable typo

leonardosepulveda committed:
change test name

leonardosepulveda committed:
simplify test json

leonardosepulveda commented:
I changed the name of the tests from test_segmentation... to test_merfish_segmentation... and the workflow works now. The only difference is that the tests run before test_plotting. Is this because the results of each analysis are in the same folder, and when trying the new analysis it reads previous analysis too? This seems to be a bug as I would not expect that the order of the execution should matter. Any ideas on how to solve it? maybe clean the folder after each test?

Now the codecov patch is the only part that is not passing. Most of the code I wrote is in segmentation.py., and has a coverage of 78%. I would need to create a 3D segmentation test to increase the coverage. Could you merge this branch to the main now or you want me to create 3D tests?

leonardosepulveda committed:
change json name in test py file

leonardosepulveda committed:
simplify json to increase codecov

leonardosepulveda committed:
100>1 in the cv2 segmentation

leonardosepulveda committed:
add flow_threshold and cellpob_threshold parameters

leonardosepulveda committed:
adding flow_thresh and cellprob to machineLearningSegment class

shavan2600 commented:
Good
